core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/provider.go

```diff
@@ -289,23 +291,36 @@ func (p *logEventProvider) getLogsFromBuffer(latestBlock int64) []ocr2keepers.Up
 	case BufferVersionV1:
 		// in v1, we use a greedy approach - we keep dequeuing logs until we reach the max results or cover the entire range.
 		blockRate, logLimitLow, maxResults, _ := p.getBufferDequeueArgs()
-		for len(payloads) < maxResults && start <= latestBlock {
-			logs, remaining := p.bufferV1.Dequeue(start, blockRate, logLimitLow, maxResults-len(payloads), DefaultUpkeepSelector)
-			if len(logs) > 0 {
-				p.lggr.Debugw("Dequeued logs", "start", start, "latestBlock", latestBlock, "logs", len(logs))
-			}
-			for _, l := range logs {
-				payload, err := p.createPayload(l.ID, l.Log)
-				if err == nil {
-					payloads = append(payloads, payload)
+
+		if p.currentIteration == 0 {
+			p.iterations = (p.bufferV1.NumOfUpkeeps() / logLimitLow) / maxResults
```

Comment:
I've wrapped this in an if because I'm assuming we want the number of iterations to stay the same for the entirety of a single "loop" of the range of iterations. This way, we calculate the total number of iterations at the start of the iterations, capturing the number of upkeeps once. The number of iterations won't change again until the full range of 0...iterations has been completed.